### PR TITLE
Adapt the Transaction::verify API to not require mutating state

### PR DESCRIPTION
**Pull request by stevenroose**
Proposal to improve a bit on #319.

I think it makes more sense for the verify method to keep track of
double-spending itself instead of requiring the closure to do that.
When internally keeping track of double-spending, the closure can just
be a simple inspector closure and doesn't need to mutate any state,
making it a lot simpler and potentially more efficient.

Also, OutPoint is Copy, so doesn't need to be referenced.

### CODE REVIEW

File: src/blockdata/transaction.rs [hunk 1]
```diff
@@ -406,15 +406,20 @@ impl Transaction {
 
     #[cfg(feature="bitcoinconsensus")]
     /// Verify that this transaction is able to spend its inputs
-    /// The lambda spent should not return the same TxOut twice!
-    pub fn verify<S>(&self, mut spent: S) -> Result<(), script::Error>
-        where S: FnMut(&OutPoint) -> Option<TxOut> {
-        let tx = serialize(&*self);
+    pub fn verify<'a, S>(&self, get_utxo: S) -> Result<(), script::Error>
+        where S: Fn(OutPoint) -> Option<&'a TxOut> {
```
Comment (apoelstra) [inline src/blockdata/transaction.rs]: Since you now call it only once, `FnOnce` would be more general and more appropriate.

Comment (stevenroose) [reply src/blockdata/transaction.rs]: I never fully understood `Fn` vs `FnOnce`. You call the method for every input, right? So not once. Unless "once" here has some extra meaning I'm missing.

Comment (elichai) [reply src/blockdata/transaction.rs]: FnOnce is a closure that *cannot* be called more than once, usually by moving a non Copy type from the outside environment to the closure.
example: https://play.rust-lang.org/?gist=9f86723ad4d2b0530a45d4047f311efc

Comment (apoelstra) [reply src/blockdata/transaction.rs]: Oh, you're right, it's called in a loop. So you need to use `FnMut`.

Please do not use `Fn`. It is never necessary and always restrictive.

**Comment by tamasblummer** [comment]
An important step of transaction verification is that it does only spend an output once, this is easy to forget as we have seen in Bitcoin Core. If you want Fn instead of FnMut then the function itself should prove that inputs are not reused. That would be better than the current code that relies that used outputs are removed.

**Comment by stevenroose** [comment]
> An important step of transaction verification is that it does only spend an output once, this is easy to forget as we have seen in Bitcoin Core. If you want Fn instead of FnMut then the function itself should prove that inputs are not reused. That would be better than the current code that relies that used outputs are removed.

This PR also does that. It checks for double spends inside the tx. For block verification or so, the verifier has to keep track of inputs on an extra level. One could argue that this method is not useful in that scenario, as then you'd be checking twice. But we learned from experience that we better check twice than accidentally not at all ;)

**Comment by tamasblummer** [comment]
maybe also no output should be below DUST limit.

**Comment by tamasblummer** [comment]
maybe also check that it meets a target fee/vByte (parameter to verify)

**Comment by stevenroose** [comment]
I fixed the `FnOnce`->`FnMut` change. But while trying to address your other comments, @tamasblummer , https://github.com/rust-bitcoin/rust-bitcoin/pull/330 made more sense to me. Otherwise we'd need a `script::Error::AmountsDontMatch` error or something and that's strange.

**Comment by tamasblummer** [comment]
yeah, let's close this one then.

**Pull request closed without merging**